protocol/signalfx/tagreplace/tagreplace_test.go

```diff
@@ -0,0 +1,93 @@
+package tagreplace
+
+import (
+	"context"
+	"github.com/signalfx/golib/datapoint"
+	"github.com/signalfx/golib/event"
+	"github.com/signalfx/golib/pointer"
+	"github.com/signalfx/golib/trace"
+	. "github.com/smartystreets/goconvey/convey"
+	"testing"
+)
+
+type end struct {
+}
+
+func (e *end) AddSpans(ctx context.Context, spans []*trace.Span) error {
+	return nil
+}
+
+func (e *end) AddDatapoints(ctx context.Context, points []*datapoint.Datapoint) error {
+	return nil
+}
+
+func (e *end) AddEvents(ctx context.Context, events []*event.Event) error {
+	return nil
+}
+
+func Test(t *testing.T) {
+	var cases = []struct {
+		desc       string
+		rules      []string
+		inputSpan  *trace.Span
+		outputSpan *trace.Span
+	}{
+		{
+			desc:       "test single replacement",
+			rules:      []string{`^\/api\/v1\/document\/(?P<documentId>.*)\/update$`},
+			inputSpan:  &trace.Span{Name: pointer.String("/api/v1/document/321083210/update")},
+			outputSpan: &trace.Span{Name: pointer.String("/api/v1/document/{documentId}/update"), Tags: map[string]string{"documentId": "321083210"}},
+		},
+		{
+			desc:       "test multi replacement",
+			rules:      []string{`^\/api\/(?P<version>.*)\/document\/(?P<documentId>.*)\/update$`},
+			inputSpan:  &trace.Span{Name: pointer.String("/api/v1/document/321083210/update")},
+			outputSpan: &trace.Span{Name: pointer.String("/api/{version}/document/{documentId}/update"), Tags: map[string]string{"documentId": "321083210", "version": "v1"}},
+		},
+	}
+	Convey("test tag replace", t, func() {
+		for _, tc := range cases {
+			Convey("Testing case: "+tc.desc, func() {
+				e := &end{}
+				tr, err := New(tc.rules, e)
+				So(err, ShouldBeNil)
+				So(tr, ShouldNotBeNil)
+				err = tr.AddSpans(context.Background(), []*trace.Span{tc.inputSpan})
+				So(err, ShouldBeNil)
+
+				So(*tc.inputSpan.Name, ShouldEqual, *tc.outputSpan.Name)
+				So(tc.inputSpan.Tags, ShouldResemble, tc.outputSpan.Tags)
+			})
+		}
+	})
+
+}
+
+func Benchmark(b *testing.B) {
+	spans := make([]*trace.Span, 0, b.N)
+	for i := 0; i < b.N; i++ {
+		spans = append(spans, &trace.Span{Name: pointer.String("/api/v1/document/321083210/update"), Tags: map[string]string{}})
+	}
+	tr, _ := New([]string{
+		`^\/api\/(?P<version>.*)\/document\/(?P<documentId>.*)\/update$`,
```

Comment:
Can you also benchmark with less greedy captures (like `(?P<version>[^\/]+)` instead of `.*`) as well and see if it's worth making recommendations there for people writing those rules?